From c4342af965b293d5a89f3cd141452c685359c568 Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Tue, 7 Oct 2025 12:54:50 +0200
Subject: [PATCH 1/9] loader: change winediag logging to Spritz

---
 dlls/ntdll/loader.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index 436d69ccac7..2496832fac4 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -4717,8 +4717,8 @@ void loader_init( CONTEXT *context, void **entry )
             }
             if (FIXME_ON(winediag))
             {
-                MESSAGE("winediag: wine-cachyos %s is a testing version containing experimental patches.\n", wine_get_version());
-                MESSAGE("winediag: this wine contains many experimental patches, please don't report bugs to winehq.org.\n");
+                MESSAGE("winediag: Spritz-Wine-CachyOS %s is a downstream version containing specific patches for games.\n", wine_get_version());
+                MESSAGE("winediag: Please don't report bugs about it on winehq.org or at wine-cachyos but use https://github.com/NelloKudo/Wine-Builds instead.\n");
             }
         }
         NtClose( startup_event );
-- 
2.52.0

