From 6d8a0189345d9cab1137e7e45df28eb09bca91fe Mon Sep 17 00:00:00 2001
From: mkrsym1 <mkrsym1@gmail.com>
Date: Mon, 8 Sep 2025 21:05:27 +0200
Subject: [PATCH 8/9] loader: Copy arial.ttf to fonts

---
 loader/wine.inf.in | 9 +++++++--
 1 file changed, 7 insertions(+), 2 deletions(-)

diff --git a/loader/wine.inf.in b/loader/wine.inf.in
index 5950fc8ffda..3ff9241c02e 100644
--- a/loader/wine.inf.in
+++ b/loader/wine.inf.in
@@ -54,7 +54,7 @@ AddService=MountMgr,0x800,MountMgrService
 RegisterDlls=RegisterDllsSection
 WineFakeDlls=FakeDllsWin32,FakeDlls
 UpdateInis=SystemIni
-CopyFiles=ColorFiles,EtcFiles,InfFiles,NlsFiles,SortFiles
+CopyFiles=ColorFiles,EtcFiles,InfFiles,NlsFiles,SortFiles,FontFiles
 AddReg=\
     Classes,\
     ContentIndex,\
@@ -169,7 +169,7 @@ AddReg=\
 [Wow64Install.ntx86]
 RegisterDlls=RegisterDllsSection
 WineFakeDlls=FakeDllsWin32,FakeDllsWow64
-CopyFiles=NlsFiles
+CopyFiles=NlsFiles,FontFiles
 AddReg=\
     CurrentVersion,\
     CurrentVersionWow64.ntx86,\
@@ -3098,9 +3098,13 @@ normnfkd.nls
 [SortFiles]
 sortdefault.nls
 
+[FontFiles]
+arial.ttf
+
 [WineSourceDirs]
 NlsFiles  = nls
 SortFiles = nls
+FontFiles = fonts
 
 [DestinationDirs]
 ColorFiles = 23
@@ -3108,6 +3112,7 @@ EtcFiles  = 12,etc
 InfFiles  = 17
 NlsFiles  = 11
 SortFiles = 10,globalization\sorting
+FontFiles = 20
 
 [NVIDIANGX]
 HKLM,Software\NVIDIA Corporation\Global\NGXCore,"FullPath",,"C:\Windows\System32"
-- 
2.51.0

