From 305b2f9349118bdd764630100dcd263faced175e Mon Sep 17 00:00:00 2001
From: NelloKudo <marshnelloosu@gmail.com>
Date: Mon, 6 Apr 2026 22:02:43 +0200
Subject: [PATCH] makedep: Small build fix.

---
 tools/makedep.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/tools/makedep.c b/tools/makedep.c
index 2acdf9cc6db..2d3329deb5f 100644
--- a/tools/makedep.c
+++ b/tools/makedep.c
@@ -4442,7 +4442,7 @@ static void output_top_makefile( struct makefile *make )
             silent_rules ? " -S" : "" );
     strarray_add( &make->phony_targets, "depend" );
 
-    if (!strarray_exists( &disabled_dirs[0], "tools/wine" ))
+    if (!strarray_exists( disabled_dirs[0], "tools/wine" ))
     {
         output( "wine: %s\n", tools_path( make, "wine" ));
         output( "\t%srm -f $@ && %s %s $@\n", cmd_prefix( "LN" ), ln_s, tools_path( make, "wine" ));
-- 
2.53.0

