From 363509f20d021fc08bc1fcda21e6df6a31b0f9f2 Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Tue, 17 Feb 2026 10:41:41 +0100
Subject: [PATCH 21/35] wininet/tests: Add tests for seeking beyond end of
 file.

---
 dlls/wininet/tests/http.c | 93 +++++++++++++++++++++++++++++++++++++++
 1 file changed, 93 insertions(+)

diff --git a/dlls/wininet/tests/http.c b/dlls/wininet/tests/http.c
index fda29394be2..df81651b942 100644
--- a/dlls/wininet/tests/http.c
+++ b/dlls/wininet/tests/http.c
@@ -587,6 +587,7 @@ static void InternetSetFilePointer_test(const char *host, const char *path)
 {
     BYTE expect_response[8192], buf[8192];
     HINTERNET hi = 0, hic = 0, hor = 0;
+    HANDLE file = INVALID_HANDLE_VALUE;
     BOOL res, expected;
     DWORD count, size, i, pos, err;
 
@@ -706,6 +707,8 @@ static void InternetSetFilePointer_test(const char *host, const char *path)
     expected = pos == INVALID_SET_FILE_POINTER && err == ERROR_INTERNET_INVALID_OPERATION;
     ok(expected, "Expected ERROR_INTERNET_INVALID_OPERATION. Got %lu\n", err);
 
+    SET_EXPECT(INTERNET_STATUS_CLOSING_CONNECTION);
+    SET_EXPECT(INTERNET_STATUS_CONNECTION_CLOSED);
     SET_EXPECT(INTERNET_STATUS_HANDLE_CLOSING);
     InternetCloseHandle(hor);
 
@@ -844,6 +847,96 @@ static void InternetSetFilePointer_test(const char *host, const char *path)
     CLEAR_NOTIFIED(INTERNET_STATUS_RECEIVING_RESPONSE);
     CLEAR_NOTIFIED(INTERNET_STATUS_RESPONSE_RECEIVED);
 
+    SET_EXPECT(INTERNET_STATUS_CLOSING_CONNECTION);
+    SET_EXPECT(INTERNET_STATUS_CONNECTION_CLOSED);
+    SET_EXPECT(INTERNET_STATUS_HANDLE_CLOSING);
+    InternetCloseHandle(hor);
+
+    SET_EXPECT(INTERNET_STATUS_HANDLE_CREATED);
+    hor = HttpOpenRequestA(hic, NULL, path, NULL, NULL, NULL,
+                           INTERNET_FLAG_RELOAD,
+                           0xdeadbead);
+    ok(hor != 0x0, "HttpOpenRequest failed: %lu\n", GetLastError());
+    if(hor == 0x0) goto abort;
+    CHECK_NOTIFIED(INTERNET_STATUS_HANDLE_CREATED);
+
+    SET_OPTIONAL(INTERNET_STATUS_CONNECTING_TO_SERVER);
+    SET_OPTIONAL(INTERNET_STATUS_CONNECTED_TO_SERVER);
+
+    SET_EXPECT(INTERNET_STATUS_SENDING_REQUEST);
+    SET_EXPECT(INTERNET_STATUS_REQUEST_SENT);
+    SET_EXPECT(INTERNET_STATUS_RECEIVING_RESPONSE);
+    SET_EXPECT(INTERNET_STATUS_RESPONSE_RECEIVED);
+
+    res = HttpSendRequestA(hor, NULL, 0, NULL, 0);
+    err = !res ? GetLastError() : NO_ERROR;
+    expected = res && err == NO_ERROR;
+    ok(expected, "HttpSendRequest failed: %lu\n", err);
+
+    CHECK_NOTIFIED(INTERNET_STATUS_SENDING_REQUEST);
+    CHECK_NOTIFIED(INTERNET_STATUS_REQUEST_SENT);
+    CHECK_NOTIFIED(INTERNET_STATUS_RECEIVING_RESPONSE);
+    CHECK_NOTIFIED(INTERNET_STATUS_RESPONSE_RECEIVED);
+
+    SetLastError(0xdeadbeef);
+    res = InternetReadFile(hor, buf, 10, &count);
+    err = !res ? GetLastError() : NO_ERROR;
+    ok(res, "InternetReadFile failed: %lu\n", err);
+    ok(count, "InternetReadFile returned no content\n");
+
+    SetLastError(0xdeadbeef);
+    pos = InternetSetFilePointer(hor, INT_MAX, NULL, FILE_BEGIN, 0);
+    err = pos == INVALID_SET_FILE_POINTER ? GetLastError() : NO_ERROR;
+    expected = pos == INT_MAX && err == NO_ERROR;
+    ok(expected, "Expected position %#x. Got %#lx. GetLastError() %lu\n", INT_MAX, pos, err);
+
+    SetLastError(0xdeadbeef);
+    res = InternetReadFile(hor, buf, 1024, &count);
+    err = !res ? GetLastError() : NO_ERROR;
+    todo_wine
+    ok(!res, "InternetReadFile succeeded unexpectedly\n");
+    todo_wine
+    ok(err == ERROR_NOACCESS, "InternetReadFile unexpected error %lu\n", err);
+
+    SetLastError(0xdeadbeef);
+    res = InternetReadFile(hor, buf, 1024, &count);
+    err = !res ? GetLastError() : NO_ERROR;
+    todo_wine
+    ok(!res, "InternetReadFile succeeded unexpectedly\n");
+    todo_wine
+    ok(err == ERROR_INTERNET_INCORRECT_HANDLE_STATE, "InternetReadFile unexpected error %lu\n", err);
+
+    SetLastError(0xdeadbeef);
+    pos = InternetSetFilePointer(hor, 0, NULL, FILE_BEGIN, 0);
+    err = pos == INVALID_SET_FILE_POINTER ? GetLastError() : NO_ERROR;
+    expected = pos == INVALID_SET_FILE_POINTER && err == ERROR_INTERNET_INVALID_OPERATION;
+    todo_wine
+    ok(expected, "Expected position %#x. Got %#lx. GetLastError() %lu\n", INVALID_SET_FILE_POINTER, pos, err);
+
+    count = 0;
+    SetLastError(0xdeadbeef);
+    res = InternetQueryDataAvailable(hor, &count, 0x0, 0x0);
+    err = !res ? GetLastError() : NO_ERROR;
+    todo_wine
+    ok(!res, "InternetQueryDataAvailable succeeded unexpectedly\n");
+    todo_wine
+    ok(err == ERROR_INTERNET_INCORRECT_HANDLE_STATE, "InternetQueryDataAvailable unexpected error %lu\n", err);
+    todo_wine
+    ok(count == 0, "InternetQueryDataAvailable unexpected count: %lu\n", count);
+
+    size = ARRAY_SIZE(buf);
+    res = InternetQueryOptionA(hor, INTERNET_OPTION_URL, buf, &size);
+    ok(res, "InternetQueryOptionA(INTERNET_OPTION_URL) failed with error %ld\n", GetLastError());
+
+    res = TRUE;
+    res = InternetSetOptionA(hor, INTERNET_OPTION_HTTP_DECODING, &res, sizeof(res));
+    ok(res, "InternetSetOption(INTERNET_OPTION_HTTP_DECODING) failed: %ld\n", GetLastError());
+
+    res = InternetLockRequestFile(hor, &file);
+    if (res)
+        InternetUnlockRequestFile(file);
+    ok(res, "InternetLockRequestFile failed: %ld\n", GetLastError());
+
 abort:
     SET_OPTIONAL(INTERNET_STATUS_CLOSING_CONNECTION);
     SET_OPTIONAL(INTERNET_STATUS_CONNECTION_CLOSED);
-- 
2.53.0

