From eaf55ef81f4d1ffbd6185a51a6929765e8164e70 Mon Sep 17 00:00:00 2001
From: Torge Matthies <openglfreak@googlemail.com>
Date: Tue, 3 Mar 2026 16:42:51 +0100
Subject: [PATCH 26/35] wininet: Do not reset content_pos in reset_data_stream.

---
 dlls/wininet/http.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/dlls/wininet/http.c b/dlls/wininet/http.c
index 57c5512a7e5..e3147e3600b 100644
--- a/dlls/wininet/http.c
+++ b/dlls/wininet/http.c
@@ -329,7 +329,6 @@ static void reset_data_stream(http_request_t *req)
     destroy_data_stream(req->data_stream);
     req->data_stream = &req->netconn_stream.data_stream;
     req->read_pos = req->read_size = req->netconn_stream.content_read = 0;
-    req->content_pos = 0;
     req->read_gzip = FALSE;
 }
 
-- 
2.53.0

